src/dataflow/core/sexp.py

```diff
@@ -42,86 +44,109 @@ def _split_respecting_quotes(s: str) -> List[str]:
-                clean_singletons=clean_singletons,
-            )
-        )
-        start = close_bracket_idx + 1
-
-    if start < len(sexp_string):
-        # tokens after the last ')'
-        postparen = sexp_string[start:].strip()
-        if postparen != "":
-            tokens = _split_respecting_quotes(postparen)
-            result.extend(tokens)
-
-    if len(result) == 2 and result[-1] == ";":
-        sexp_tmp = result[0]
-    else:
-        sexp_tmp = result
-
-    # special-case top-level values because they need an extra level
-    # of parens in the Sexp:
-    if isinstance(result, list) and len(result) >= 1 and result[0] == "#":
-        return [result]
-    if clean_singletons and len(sexp_tmp) == 1:
-        return sexp_tmp[0]
-    return sexp_tmp
+def parse_sexp(s: str) -> Sexp:
+    offset = 0
+
+    # eoi = end of input
+    def is_eoi():
+        nonlocal offset
+        return offset == len(s)
+
+    def peek():
+        nonlocal offset
+        return s[offset]
+
+    def next_char():
+        # pylint: disable=used-before-assignment
+        nonlocal offset
+        cn = s[offset]
+        offset += 1
+        return cn
+
+    def skip_whitespace():
+        while (not is_eoi()) and peek().isspace():
+            next_char()
+
+    def skip_then_peek():
+        skip_whitespace()
+        return peek()
+
+    def read() -> Sexp:
+        skip_whitespace()
+        c = next_char()
+        if c == LEFT_PAREN:
+            return read_list()
+        elif c == DOUBLE_QUOTE:
+            return read_string()
+        elif c == META:
+            meta = read()
+            expr = read()
+            return [META, meta, expr]
+        elif c == READER:
+            return [READER, read()]
+        else:
+            out_inner = ""
+            if c != "\\":
+                out_inner += c
+
+            # TODO: is there a better loop idiom here?
+            if not is_eoi():
+                next_c = peek()
+                escaped = c == "\\"
+                while (not is_eoi()) and (
+                    escaped or not _is_beginning_control_char(next_c)
+                ):
+                    if (not escaped) and next_c == "\\":
+                        next_char()
+                        escaped = True
+                    else:
+                        out_inner += next_char()
+                        escaped = False
+                    if not is_eoi():
+                        next_c = peek()
+            return out_inner
+
+    def read_list():
+        out_list = []
+        while skip_then_peek() != RIGHT_PAREN:
+            out_list.append(read())
+        next_char()
+        return out_list
+
+    def read_string():
+        out_str = ""
+        while peek() != '"':
+            c_string = next_char()
+            out_str += c_string
+            if c_string == "\\":
+                out_str += next_char()
+        next_char()
+        return f'"{out_str}"'
+
+    out = read()
+    skip_whitespace()
```

Comment:
this looks like it doesn't do anything?

Reply:
Thanks, there's a missing assertion that parsing has completed.